### PR TITLE
GH-15993: Custom metric as a hyperparam for grid search

### PR DESCRIPTION
**Pull request by tomasfryda**
#15993 

I encountered the following error several times but the next day I wasn't able to reproduce it but I'm looking into it now.
```java
01-02 17:55:27.689 192.168.0.73:54321    9334       FJ-2-7 ERROR water.default:
org.python.core.PyException: ImportError: reload(): module mm_mae not in sys.modules
	at org.python.core.Py.ImportError(Py.java:327) ~[h2o.jar:?]
	at org.python.core.imp._reload(imp.java:1751) ~[h2o.jar:?]
	at org.python.core.imp.reload(imp.java:1739) ~[h2o.jar:?]
	at org.python.core.__builtin__.reload(__builtin__.java:1079) ~[h2o.jar:?]
	at org.python.core.__builtin__.reload(__builtin__.java:1075) ~[h2o.jar:?]
	at water.udf.JythonObjectFactory.<init>(JythonObjectFactory.java:21) ~[h2o.jar:?]
	at water.udf.JythonCFuncLoader.load(JythonCFuncLoader.java:36) ~[h2o.jar:?]
	at water.udf.CFuncTask.setupLocal(CFuncTask.java:23) ~[h2o.jar:?]
	at water.MRTask.setupLocal0(MRTask.java:734) ~[h2o.jar:?]
	at water.MRTask.dfork(MRTask.java:628) ~[h2o.jar:?]
	at water.MRTask.doAll(MRTask.java:529) ~[h2o.jar:?]
	at water.MRTask.doAll(MRTask.java:470) ~[h2o.jar:?]
	at hex.tree.Score.scoreAndMakeModelMetrics(Score.java:170) ~[h2o.jar:?]
	at hex.tree.SharedTree.doScoringAndSaveModel(SharedTree.java:851) ~[h2o.jar:?]
	at hex.tree.SharedTree$Driver.scoreAndBuildTrees(SharedTree.java:488) ~[h2o.jar:?]
	at hex.tree.SharedTree$Driver.computeImpl(SharedTree.java:405) ~[h2o.jar:?]
	at hex.ModelBuilder$Driver.compute2(ModelBuilder.java:253) ~[h2o.jar:?]
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1689) ~[h2o.jar:?]
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468) ~[h2o.jar:?]
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263) [h2o.jar:?]
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976) [h2o.jar:?]
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479) [h2o.jar:?]
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104) [h2o.jar:?]
```

### CODE REVIEW

**Comment by wendycwong** [comment]
@tomasfryda :  I am thinking about this:

When you use gridsearch, after you got the models back, you can call grid_model.sort_by(metric="my_metric") to sort your grid search model and choose the model with the best value listed in my_metric.  Right now we only accept certain metrics and not custom metric names.  I was wondering if you can just make sure we accept custom_metrics names in the sort_by methods in Python and R and call it done.

Second, there is also this thing of using custom metrics as an early stopping criteria metrics.  I don't think we are doing this right now.  So, I think perhaps we can just do the above and forget about this part for now? 

Customer claims that you are on a deadline so I think we can just get them the first part is good enough.

Thanks, W

File: h2o-docs/src/product/data-science/coxph.rst [hunk 1]
```diff
@@ -69,7 +69,7 @@ Common parameters
   
     **NOTE**: In Flow, if you click the **Build a model** button from the ``Parse`` cell, the training frame is entered automatically.
 
-- `use_all_factor_levels <algo-params/coxph.html>`__: Specify whether to use all factor levels in the possible set of predictors; if you enable this option, sufficient regularization is required. By default, the first factor level is skipped. This option defaults to ``True`` (enabled).
+- `use_all_factor_levels <algo-params/use_all_factor_levels.html>`__: Specify whether to use all factor levels in the possible set of predictors; if you enable this option, sufficient regularization is required. By default, the first factor level is skipped. This option defaults to ``True`` (enabled).
```
Comment (tomasfryda) [inline h2o-docs/src/product/data-science/coxph.rst:72]: Unrelated to this PR. I accidentally noticed this error when looking for `custom_metric_func` documentation.

File: h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst [hunk 2]
```diff
@@ -3,7 +3,7 @@
 ``custom_metric_func``
 ----------------------
 
-- Available in: GBM, DRF, Deeplearning, Stacked Ensembles, GLM
+- Available in: GAM, GBM, GLM, DRF, Deeplearning, Stacked Ensembles
```
Comment (tomasfryda) [inline h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst:6]: @wendycwong could you confirm that GLM and GAM support `custom_metric_func`? From the code it looks like it should be supported.

**Comment by tomasfryda** [comment]
> When you use gridsearch, after you got the models back, you can call grid_model.get_grid(sort_by="my_metric") to sort your grid search model and choose the model with the best value listed in my_metric. Right now we only accept certain metrics and not custom metric names. I was wondering if you can just make sure we accept custom_metrics names in the get_grid methods in Python and R and call it done.

@wendycwong we accept the metrics through reflection on model metrics object. I'd keep that but in future we might refactor AutoML and use the smart behavior from there to use the metrics and decide how it should be sorted.

I added the possibility to use `custom` and `custom_increasing`, unfortunately grid search is not smart about metrics so it assumes that every specified metric is a loss (lower is better) so `custom` and `custom_increasing` behaves the same and `custom_increasing` is there just as a convenience for the user (e.g. so you can sort by the same metric as you use in stopping metric). You can however specify `gs.get_grid(sort_by="custom_increasing", decreasing=True)` and you get the right behavior.

If you meant using the custom_metric name (`CFuncRef.keyName` or `CFuncRef.funcName`) as the name, I would not recommend that as it would be harder and wouldn't provide any real advantages - you could sort by `My_custom_metric` but if you'd use more than one you wouldn't have `My_custom_metric` defined for all the models and the problem with `get_grid(sort_by=...)` would persist.

> Second, there is also this thing of using custom metrics as an early stopping criteria metrics. I don't think we are doing this right now. So, I think perhaps we can just do the above and forget about this part for now?

That's already enabled if the user specifies the `stopping_metric`. The behavior is following (see the tests for more):

```python
    hyper_parameters["ntrees"] = ntrees_opts
    hyper_parameters["custom_metric_func"] = [custom_mae_mm(), custom_rmse_mm()]
    hyper_parameters["stopping_metric"] = "custom"
    ...
    gs = H2OGridSearch(H2OGradientBoostingEstimator, hyper_params=hyper_parameters)
    ...
    print(gs.get_grid(sort_by="rmse"))
```

```
Hyper-Parameter Search Summary: ordered by increasing rmse
    custom_metric_func                                   ntrees    stopping_metric    model_ids                                                        rmse
--  ---------------------------------------------------  --------  -----------------  ---------------------------------------------------------------  --------
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  10        custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_8  0.310892
    python:mae=mm_mae.CustomMaeFuncWrapper               10        custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_7  0.315669
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  5         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_6  0.475182
    python:mae=mm_mae.CustomMaeFuncWrapper               5         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_5  0.480646
    python:mae=mm_mae.CustomMaeFuncWrapper               3         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_3  0.571315
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  3         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_4  0.572644
    python:mae=mm_mae.CustomMaeFuncWrapper               1         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_1  0.689742
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  1         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_2  0.69047
```

```python
    print(gs.get_grid(sort_by="mae"))
```

```
Hyper-Parameter Search Summary: ordered by increasing mae
    custom_metric_func                                   ntrees    stopping_metric    model_ids                                                        mae
--  ---------------------------------------------------  --------  -----------------  ---------------------------------------------------------------  --------
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  10        custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_8  0.259908
    python:mae=mm_mae.CustomMaeFuncWrapper               10        custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_7  0.267279
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  5         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_6  0.404209
    python:mae=mm_mae.CustomMaeFuncWrapper               5         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_5  0.407375
    python:mae=mm_mae.CustomMaeFuncWrapper               3         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_3  0.489523
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  3         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_4  0.491984
    python:mae=mm_mae.CustomMaeFuncWrapper               1         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_1  0.596325
    python:rmse=mm_rmse.CustomRmseRegressionFuncWrapper  1         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_14_model_2  0.598226
```

```python
        print(gs.get_grid(sort_by="custom"))
```

This fails since there is multiple custom metric functions.
```
h2o.exceptions.H2OResponseError: Server error water.exceptions.H2OIllegalArgumentException:
  Error: Invalid argument for sort_by specified. Using multiple definitions of custom metric.
  Request: GET /99/Grids/Grid_GBM_iris_wheader.hex_model_python_1704361943199_26
    params: {'sort_by': 'custom'}
```

But if there is just one custom metric it will succeed and sort according to custom metric
```
Hyper-Parameter Search Summary: ordered by increasing custom
    custom_metric_func                      ntrees    stopping_metric    model_ids                                                        custom
--  --------------------------------------  --------  -----------------  ---------------------------------------------------------------  --------
    python:mae=mm_mae.CustomMaeFuncWrapper  10        custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_29_model_4  0.264247
    python:mae=mm_mae.CustomMaeFuncWrapper  5         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_29_model_3  0.407098
    python:mae=mm_mae.CustomMaeFuncWrapper  3         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_29_model_2  0.49106
    python:mae=mm_mae.CustomMaeFuncWrapper  1         custom             Grid_GBM_iris_wheader.hex_model_python_1704361943199_29_model_1  0.60098
```
or custom_increasing
```
Hyper-Parameter Search Summary: ordered by decreasing custom_increasing
    custom_metric_func                                             ntrees    stopping_metric    model_ids                                                        custom_increasing
--  -------------------------------------------------------------  --------  -----------------  ---------------------------------------------------------------  -------------------
    python:nrmse=mm_nrmse.CustomNegativeRmseRegressionFuncWrapper  10        custom_increasing  Grid_GBM_iris_wheader.hex_model_python_1704361943199_27_model_4  -0.31107
    python:nrmse=mm_nrmse.CustomNegativeRmseRegressionFuncWrapper  5         custom_increasing  Grid_GBM_iris_wheader.hex_model_python_1704361943199_27_model_3  -0.47747
    python:nrmse=mm_nrmse.CustomNegativeRmseRegressionFuncWrapper  3         custom_increasing  Grid_GBM_iris_wheader.hex_model_python_1704361943199_27_model_2  -0.57942
    python:nrmse=mm_nrmse.CustomNegativeRmseRegressionFuncWrapper  1         custom_increasing  Grid_GBM_iris_wheader.hex_model_python_1704361943199_27_model_1  -0.691999

```

Please @wendycwong  let me know if it resolves your suggestions.  Thank you!

> Customer claims that you are on a deadline so I think we can just get them the first part is good enough.

I wasn't aware of that but the customer is always right so I'll try to finish ASAP.

File: h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst (hunk 2 above)
Comment (wendycwong) [reply h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst:6]: Yuliia added support for custom_metric_func for GLM.  GAM does not support custom_metric_func yet.

**Comment by wendycwong** [comment]
Tomas:  

I meant to say the customer is on some sort of deadline.  Sorry for the confusion. 

W

File: h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst (hunk 2 above)
Comment (wendycwong) [reply h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst:6]: Okay, there is no test on it so I will treat it as it is not there.  Opened an issue here to make sure it works: https://github.com/h2oai/h2o-3/issues/16006 .  For now please treat it as it is not supported.

**Review by wendycwong — CHANGES_REQUESTED** [review]

**Comment by wendycwong** [comment]
@tomasfryda 

I think I finally know what my confusion is about now.

Normally, the hyperparameters are just parameters and it may affect the model metrics to various degree.

However, if we let custom_metric to be a hyperparameters, they are both parameters and model metrics.  In this case, if you allow custom_metric to be griddable say we have custom_rmse and custom_mae.  Then, in one set of model, the custom_rmse will be included in its metrics.  In another set of model, the custom_mae will be in metrics.  Assume for the moment that your get_grid model allow you to set sort_by to any custom metric names you want.  In this case, if you do gs.get_grid(sort_by="custom_rmse"), you will run into errors because half of the model will have custom metrics custom_rmse, the other half will have custom_mae and not custom_rmse.  So due to the absence of one custom_metric, you will run into error.

On the other hand, if you just specify sort_by = "custom" and force the get_grid to just grab and sort according to whatever custom metrics the model have, it still does not make sense to sort all models on either particular custom metrics because it may not make sense to compare the two custom metrics.

So, my conclusion is not to make custom_metric to be a hyperparameter.  

You already have completed the hard part, i.e. make them stopping metrics.  To avoid the whole confusion of making them hyperparameter but that will fail when you try to call get_grid() to sort your models.  

I think it is more important to be able to sort the grid models by the custom_metric values than making it a hyperparameter. 

What do you think? Tomas?  If you agree, this PR is mostly done!

**Comment by tomasfryda** [comment]
@wendycwong I think we are still confused. From my view there are 2 use-cases:
1) get custom metric calculated on the final model and CV models
2) use custom metrics as a hyper-parameter to influence the training (as `stopping_metric`)

The first is already possible when you keep the cv predictions and cv folds - you can calculate the metric manually in python/R. The second is currently not possible.

But it's important to note that the current PR does not make a choice between the two - it supports both.

If you want to do the first use-case you specify just one custom metric as hyper-param - this way there is just one custom metric and `get_grid(sort_by="custom")` works as expected.

When you want the second use-case you can just specify more custom metrics and set stopping_metric="custom". This way you can choose metrics that will e.g. weight differently outliers. This has the downside that there are multiple definitions of custom metric and thus it makes no sense to do `get_grid(sort_by="custom")` but you can still get the models sorted by some other common metrics (e.g. MAE, RMSE).

For now I removed the possibility to have `custom_metric_func` as hyperparameter as you wanted, so the usage is:

```python
gs = H2OGridSearch(H2OGradientBoostingEstimator(custom_metric_func=custom_nrmse_mm()),
                       hyper_params=hyper_parameters)
gs.train(...)
print(gs.get_grid(sort_by="custom"))
```

This change is isolated in one commit so I can revert to previous behavior easily.

File: h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomasfryda) [reply h2o-docs/src/product/data-science/algo-params/custom_metric_func.rst]: Ok, I removed it from the docs and also i removed it from GAM schema so it doesn't appear in the API and h2o flow until it's tested. Is that OK @wendycwong  or do you prefer to have it enabled but not documented?

**Comment by wendycwong** [comment]
@tomasfryda:

I feel things are pretty clear right now.  You can use custom_metric as a stopping metric and you can sort your grid models/autoML models using custom_metrics.  This is pretty cool.

Thanks,
Wendy

**Review by wendycwong — DISMISSED** [review]

**Comment by tomasfryda** [comment]
Sorry @wendycwong, I accidentally pushed here a commit that should have end up in the test image PR so I reverted it but at the same time lost your approval. Could you please approve this PR again?

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Pull request merged**